src/main/java/minijava/ir/assembler/SimpleNodeAllocator.java

```diff
@@ -70,25 +64,28 @@ public Location getLocation(Node node) {
       if (proj.getPred().equals(graph.getArgs())) {
         // the proj node points to a method argument
         int slot = proj.getNum();
-        if (useABIConformCalling) {
-          if (slot < Register.methodArgumentQuadRegisters.size()) {
-            return regPassedParameterLocations.get(slot);
-          }
-          slot = slot - Register.methodArgumentQuadRegisters.size();
-        }
-        // this seems to be correct?
-        return new RegRelativeLocation(Register.BASE_POINTER, (slot + 2) * STACK_SLOT_SIZE);
+        return paramLocations.get(slot);
       } else if (proj.getPred().getPredCount() > 0 && proj.getPred().getPred(0) instanceof Call) {
         // the proj node points to method result
         // the result always lays in the RAX register
         // [Call] <- [Tuple Proj] <- [Is Proj]
-        return getStackSlotAsLocation(proj.getPred().getPred(0));
+        return getNodeLocation(proj.getPred().getPred(0));
       } else if (node.getPred(0) instanceof Load || node.getPred(0) instanceof Store) {
         // [Proj p64] <- [Load|Store] <- [Proj res]
         return getLocation(node.getPred(0));
       }
     }
-    return getStackSlotAsLocation(node);
+    if (node instanceof firm.nodes.Call) {
+      return getNodeLocation(
+          node,
+          ((MethodType) (((Address) node.getPred(1)).getEntity().getType()))
+              .getResType(0)
+              .getMode());
+    }
+    if (node instanceof firm.nodes.Load) {
```

Comment:
Visitor is most useful for making sure not to have forgotten a case at compile time. If you aren't matching all cases anyway, using a visitor just adds noise IMO.